pkg/k8s/watchers/cilium_network_policy.go

```diff
@@ -388,3 +566,130 @@ func reportCNPChangeMetrics(err error) {
 		metrics.PolicyChangeTotal.WithLabelValues(metrics.LabelValueOutcomeSuccess).Inc()
 	}
 }
+
+func resolveCIDRGroupRef(cnp *types.SlimCNP, cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup) *types.SlimCNP {
+	refs := getCIDRGroupRefs(cnp)
+	if len(refs) == 0 {
+		return cnp
+	}
+
+	cidrsSets, err := cidrGroupRefsToCIDRsSets(refs, cidrGroupCache)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+			logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			"cidrGroupRefs":                   refs,
+		}).WithError(err).Warning("unable to translate all cidr groups to cidrs")
+	}
+	translated := translateFromCIDRGroupRefs(cnp, cidrsSets)
+
+	// If all referenced CiliumCIDRGroups are empty, we need to check if FromCIDR
+	// is empty. In that case, the policy will deny all traffic, so we set the field to nil.
+	if translated.Spec != nil {
+		for i := range translated.Spec.Ingress {
+			if len(translated.Spec.Ingress[i].FromCIDR) == 0 {
+				translated.Spec.Ingress[i].FromCIDR = nil
+			}
+		}
+	}
+	for i := range translated.Specs {
+		for j := range translated.Specs[i].Ingress {
+			if len(translated.Specs[i].Ingress[j].FromCIDR) == 0 {
+				translated.Specs[i].Ingress[j].FromCIDR = nil
+			}
+		}
+	}
+	return translated
+}
+
+func hasCIDRGroupRef(cnp *types.SlimCNP, cidrGroup string) bool {
+	if specHasCIDRGroupRef(cnp.Spec, cidrGroup) {
+		return true
+	}
+	for _, spec := range cnp.Specs {
+		if specHasCIDRGroupRef(spec, cidrGroup) {
+			return true
+		}
+	}
+	return false
+}
+
+func specHasCIDRGroupRef(spec *api.Rule, cidrGroup string) bool {
+	if spec == nil {
+		return false
+	}
+	for _, ingress := range spec.Ingress {
+		for _, cidrGroupRef := range ingress.FromCIDRGroupRefs {
+			if string(cidrGroupRef) == cidrGroup {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getCIDRGroupRefs(cnp *types.SlimCNP) []string {
+	var cidrGroupRefs []string
+
+	if cnp.Spec != nil {
+		for _, ingress := range cnp.Spec.Ingress {
+			for _, cidrGroupRef := range ingress.FromCIDRGroupRefs {
+				cidrGroupRefs = append(cidrGroupRefs, string(cidrGroupRef))
+			}
+		}
+	}
+	for _, spec := range cnp.Specs {
+		for _, ingress := range spec.Ingress {
+			for _, cidrGroupRef := range ingress.FromCIDRGroupRefs {
+				cidrGroupRefs = append(cidrGroupRefs, string(cidrGroupRef))
+			}
+		}
+	}
+
+	return cidrGroupRefs
+}
+
+func translateFromCIDRGroupRefs(cnp *types.SlimCNP, cidrsSets map[string][]api.CIDR) *types.SlimCNP {
+	cnpCpy := cnp.DeepCopy()
```

Comment:
I think we do to keep the "untranslated" cnp in the cache. See [this comment](https://github.com/cilium/cilium/pull/24638#discussion_r1154971360).

Reply:
Ah makes sense. In that case, can we move the `DeepCopy` call to when it's actually needed (as in if we need to perform translation). We should prevent performing a copy of the CNP in the general case (when user is not using CIDRGroupsRefs).

Reply:
Actually, it should already be like that.

In case of policy update, we call `translateFromCIDRGroupRefs` only after checking that the cnp has a reference to at least one CIDRGroup in `resolveCIDRGroupRef`:
```
	refs := getCIDRGroupRefs(cnp)
	if len(refs) == 0 {
		return cnp
	}
```

otherwise we return the cnp itself with no further copy.
In case of an update or a deletion of a referenced CIDRGroup, we call `resolveCIDRGroupRef` only after checking that the current cnp actually has a reference to the updated CIDRGroup:

```
		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
			continue
		}
```

Maybe here we can improve a bit, because after this we already know that the cnp has a reference to the updated CIDRGroup, so there's no reason to check for refs at the start of `resolveCIDRGroupRef` once again.

In case of a reference being present, we need the double copy because:
- the cache needs to contain the "original" cnp ahead of translation
- `k.updateCiliumNetworkPolicyV2`might modify the `oldRuleCpy` passed in, since the method `.Parse()` it is called on it and that triggers a sanitization of the rules specs: (`*IngressCommonRule) SetAggregatedSelectors()` is an example of a potential write due to the sanitization.